docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,552 @@
+            // 11. Upsert resource condition 'LateInitialized=True'.
+
+            {{- if $hookCode := Hook .CRD "post_set_late_initialized_defaults" }}
+                {{ $hookCode }}
+            {{- end }}
+            // return latest with late initialized fields updates 
+        {{- end }}
+    }
+    ```
+
+### ACK Runtime Changes
+
+* Current reconciler source code
+    ```go
+    func (r *reconciler) Sync() {
+        var latest acktypes.AWSResource
+        latest, err := rm.ReadOne()
+
+        if err!= nil {
+            if err != NotFound {
+                patchResourceStatus()
+                return err
+            }
+            latest,err := rm.Create()
+            if err != nil {
+                patchResourceStatus()
+                return err
+            }
+        } else {
+            delta := r.rd.Delta(desired, latest)
+            If delta.DifferentAt("Spec") {
+                latest,err := rm.Update()
+                if err != nil {
+                    patchResourceStatus()
+                    return err
+                }
+            }
+        }
+
+        r.patchResource(ctx, desired, latest)
+
+        // iterate through latest.conditions for requeue
+
+        return nil
+    }
+    ```
+
+* Proposed reconciler source code
+
+    ```go
+    func (r *reconciler) Sync() {
+        observed, err := rm.ReadOne()
+        var latest acktypes.AWSResource
+
+        if err!= nil {
+            if err != NotFound {
+                patchResourceStatus()
+                return err
+            }
+            // rm.Create() also updates the server side defaults from CreateResponse
+            latest,err := rm.Create()
+            if err != nil {
+                patchResourceStatus()
+                return err
+            }
+        } else {
+            delta := r.rd.Delta(desired, observed)
+            If delta.DifferentAt("Spec") {
+                // rm.Update() also updates the server side defaults from UpdateResponse
+                latest,err := rm.Update()
+                if err != nil {
+                    patchResourceStatus()
+                    return err
+                }
+            }
+        }
+
+        // updates the latest object with server side defaults.
+        // This operation supports hooks for service team to use.
+        // In case of no late initialization, lateInitializedLatest and 
+        // latest will be same.
+        lateInitializedLatest, lateInitializeRequeueErr := rm.SetLateInitializedDefaults(ctx, latest)
+
+        // Always persist the object state before exiting.
+        // patchResource will only Patch, if there is a diff.
+        if lateInitializedLatest != nil {
+            r.patchResource(ctx, desired, lateInitializedLatest) // patches metadata, spec and status if needed
+            if lateInitializeRequeueErr != nil {
+                return lateIntializedRequeueErr
+            }
+        // iterate through latest.conditions for requeue
```

Comment:
should this be one line below? i.e. after the close of the `if` block?

Reply:
I think this is at the correct place.

if lateInitializedLatest is not nil, 
a) patch the resource completely 
b) If rm.LateInitialize() requested a requeue, perform requeue 
c) Parse through conditions and determine requeue based for ConditionTypeResourceSynced